Yarn: Set 'globalFolder' in override YarnRc && drop _fetch_dependencies helper

eskultety opened the pull request:
# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- ~[ ] Docs updated (if applicable)~
- ~[ ] Docs links in the code are still valid (if docs were updated)~

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

slimreaper35 reviewed (approved):
:+1:

brunoapimentel commented:
Good catch on the `YARN_GLOBAL_FOLDER`, I did not notice it there before. We should rely on `.yarnrc.yml` to set the config we need.

brunoapimentel commented on tests/unit/package_managers/yarn/test_main.py:
```diff
@@ -103,25 +102,6 @@ def test_configure_yarn_version_fail(
-        _fetch_dependencies(source_dir, output_dir)
-
-    mock_yarn_cmd.assert_called_once_with(
-        ["install", "--mode", "skip-build"],
```
One of the things this test was checking was that `run_yarn_cmd` was being called with the right parameters. I think this could be moved to `test_resolve_yarn_project`, but this test does not exist currently.

We should do a follow up PR to increase the coverage on the yarn/main module: https://coveralls.io/builds/64062886/source?filename=cachi2%2Fcore%2Fpackage_managers%2Fyarn%2Fmain.py#L51

eskultety replied:
There's a fine line where the code coverage is a nice thing and where it leads to duplicate test cases IMO. At first I didn't see your point, but now I see you're specifically interested in testing a given set of call parameters which wouldn't be tested anywhere anymore. I'll simplify the current test and keep it around instead of removing it - we can schedule it for a conversion to the `test_resolve_yarn_project` as you noted in a follow up PR. However, as I said, we need to be careful with the test cases we test, since in this particular case we're talking about higher level helpers which transitively exercise functionality of the elementary helper primitives for which we have like 99.8% coverage of various inputs, so we'd only be creating redundancy with an extensive parametrization of say `_fetch_dependencies` or future `test_resolve_yarn_project`. Anyways, for now, I'll keep things as they are :).

brunoapimentel commented on cachi2/core/package_managers/yarn/main.py:
```diff
@@ -133,19 +134,6 @@ def _check_yarn_cache(source_dir: RootedPath) -> None:
-
-    :param source_dir: the directory in which the yarn command will be called.
-    :param output_dir: the directory where the yarn dependencies will be downloaded to.
-    :raises YarnCommandError: if the 'yarn install' command fails.
```
We should probably move this "raises" docstring to the `_resolve_yarn_project` function.

eskultety replied:
Since I kept `_fetch_dependencies` around what I'll do is that I'll also copy this raises docs note to `_resolve_yarn_project` as well.

brunoapimentel commented on cachi2/core/package_managers/yarn/main.py:
```diff
@@ -47,7 +47,8 @@ def _resolve_yarn_project(project: Project, output_dir: RootedPath) -> list[Comp
             _check_yarn_cache(project.source_dir)
         else:
-            _fetch_dependencies(project.source_dir, output_dir)
+            # fetch the dependencies
```
I like that we simplified (and fixed) the code with these changes, but I don't like that we lost the context that the function `_fetch_dependencies` brought, and now we rely on a comment to convey it.

I don't have a good suggestion to fix it though, so this is just me nitpicking :shrug:

eskultety replied:
I can put it back, I don't have a preference really; since it's quite common practice to have one-liner instance methods just to convey semantic meaning, despite me not liking the same for regular functions, sure, why not, I don't really have a good argument against such practice (I'm afraid unnecessary calls and stack operations wouldn't hold and would also be inconsistent with the practice for methods anyway) :)

brunoapimentel commented:
It would be good to have this rebased, as the new changes in main affect the content of `_resolve_yarn_project`.

eskultety commented:
> It would be good to have this rebased, as the new changes in main affect the content of `_resolve_yarn_project`.

Yeah, this is always an issue when creating new PRs - do you base it and depend on another PR or you start fresh from main? Does one need to rebase the series frequently without a review or not? Either way you'll end up with merge conflicts one will have to solve at some point :). Unless the code becomes completely irrelevant wrt/ to the latest main I usually go with the approach to gather ideas and comments on v1 and proceed with a rebase in next revisions.

eskultety commented:
since v1:
- rebased on top of latest main HEAD
- simplified and kept `_fetch_dependencies` instead of removing it
- simplified and kept `test_fetch_dependencies` instead of removing it
